core/src/apps/common/passphrase.py

```diff
@@ -41,6 +38,9 @@ async def _request_from_user(ctx: wire.Context) -> str:
 
 
 async def _request_on_host(ctx: wire.Context) -> str:
+    from trezor.messages.PassphraseAck import PassphraseAck
+    from trezor.messages.PassphraseRequest import PassphraseRequest
```

Comment:
Are there any guidelines on when to import at the module level and when to import in a function like this?

Reply:
well that would be something good to document, right?

In general, "avoid importing big modules unless you know that everyone who imports this file either (a) already has them or (b) actually needs them".

It seems `passphrase.py` is done wrong, too; hiding `PassphraseRequest` and `PassphraseAck` is probably not necessary, but importing `ui` at top level is not a good thing -- `apps.common.passphrase` is pulled in from somewhere important, but the user might not even be using passphrase, or the current session might have it cached, etc.

The most important general pattern is avoiding importing `ui` on top level in files that could reasonably be used in a silent workflow.

Reply:
Yeah, making this part of documentation would certainly decrease the odds of someone like me accidentally making the fragmentation worse again.

Reply:
imports in `apps.common.passphrase` fixed in 3aa8fe4

docs still TBD